File: server/utils/analytics/fergie.ts
import { scorePredictions, type PredictionInput } from '../scoring/engine'
import { outcomeOf, type Outcome, type Scoreline } from '../scoring/tiers'
import { minuteValue } from '../stats/insights'
import type { ScoringRules } from '../scoring/config'
import type { FergieMatch, FergieTime } from '#shared/types/analytics'

// minuteValue's sentinel for a null/unparseable minute; a real clock never
// reaches it (max ~130'), so >= this means the timeline order is unknown.
const UNKNOWN_MINUTE = 1e9

export interface FergieGoal {
  side: 'HOME' | 'AWAY'
  // Free-text upstream minute: "66'", "90'+5'", "45'+2'" (added time carries a "+").
  minute: string | null
}

// One picked, finished match fed to the Fergie replay: the user's pick plus the
// whole locked field for that match (so the rarity bonus can be re-priced at any
// hypothetical scoreline), the goal timeline, and how the match was scored.
export interface FergieMatchInput {
  home: string
  away: string
  homeCode: string | null
  awayCode: string | null
  // The user's own prediction id - it sits inside `field`, and its scored total
  // at the actual scoreline equals the stored prediction.totalPoints.
  predId: string
  // The user's pick, for the breakdown label (predId points at the same row in
  // `field`, but carrying it avoids a lookup and a can't-happen fallback).
  pred: Scoreline
  isJoker: boolean
  actual: Scoreline
  // Knockout stages force the joker multiplier on everyone (finalize's
  // countsDouble(stage)); mirror it so the replayed points match reality.
  forceJoker: boolean
  field: PredictionInput[]
  goals: FergieGoal[]
  // Closing odds of a given outcome, needed only when the config scores on ODDS.
  oddsForOutcome?: (outcome: Outcome) => number | null
}

export function emptyFergie(): FergieTime {
  return {
    matches: 0,
    goals: 0,
    netPoints: 0,
    pointsWon: 0,
    pointsLost: 0,
    biggestGain: null,
    biggestLoss: null,
    breakdown: [],
  }
}

// "Fergie time" is end-of-half stoppage from the second half on: an added-time
// goal (minute carries a "+") whose base minute is 90' or later (90'+, 105'+,
// 120'+). A first-half "45'+2'" is stoppage too, but its marginal value against a
// mid-match scoreline is not what the metric is about, so it does not count.
// minuteValue encodes base*100 + stoppage, so its hundreds are the base minute.
export function isAddedTime(minute: string | null): boolean {
  return !!minute && minute.includes('+') && Math.floor(minuteValue(minute) / 100) >= 90
}

function realPoints(m: FergieMatchInput, rules: ScoringRules, at: Scoreline): number {
  const scores = scorePredictions({
    actual: at,
    predictions: m.field,
    rules,
    actualOutcomeOdds: m.oddsForOutcome?.(outcomeOf(at)) ?? null,
    forceJoker: m.forceJoker,
  })
  return scores.find((s) => s.id === m.predId)?.totalPoints ?? 0
}

// Replay one match goal by goal, banking each added-time goal's real-points
// delta. Returns null when the timeline cannot be trusted - a goal with an
// unparseable/absent minute (so the order is unknown), or a set of goals that
// does not reconcile with the full-time score - or when no added-time goal fell.
// Bailing out here means an incomplete feed never invents a swing.
function replayMatch(m: FergieMatchInput, rules: ScoringRules): FergieMatch | null {
  const ordered = m.goals.map((g) => ({ g, order: minuteValue(g.minute) }))
  // One unorderable minute makes every added-time delta suspect (a real early
  // goal could be applied after a late one), even if the final still reconciles.
  if (ordered.some((o) => o.order >= UNKNOWN_MINUTE)) return null
  const sorted = ordered.sort((a, b) => a.order - b.order).map((o) => o.g)
  let home = 0
  let away = 0
  let gained = 0
  let lost = 0
  let added = 0
  for (const g of sorted) {
    const before: Scoreline = { home, away }
    if (g.side === 'HOME') home += 1
    else away += 1
    if (!isAddedTime(g.minute)) continue
    added += 1
    const delta = realPoints(m, rules, { home, away }) - realPoints(m, rules, before)
    if (delta > 0) gained += delta
    else if (delta < 0) lost += -delta
  }
  if (home !== m.actual.home || away !== m.actual.away) return null
  if (added === 0) return null
  return {
    home: m.home,
    away: m.away,
    homeCode: m.homeCode,
    awayCode: m.awayCode,
    predicted: `${m.pred.home}-${m.pred.away}`,
    actual: `${m.actual.home}-${m.actual.away}`,
    gained,
    lost,
    net: gained - lost,
    isJoker: m.isJoker,
  }
}

export function computeFergie(matches: FergieMatchInput[], rules: ScoringRules): FergieTime {
  const result = emptyFergie()
  for (const m of matches) {
    const replay = replayMatch(m, rules)
    if (!replay) continue
    result.matches += 1
    result.goals += m.goals.filter((g) => isAddedTime(g.minute)).length
    result.pointsWon += replay.gained
    result.pointsLost += replay.lost
    if (replay.gained > 0 && (!result.biggestGain || replay.gained > result.biggestGain.gained)) {
      result.biggestGain = replay
    }
    if (replay.lost > 0 && (!result.biggestLoss || replay.lost > result.biggestLoss.lost)) {
      result.biggestLoss = replay
    }
    // Only matches whose points actually moved make the breakdown; a match with
    // an added-time goal that changed nothing (net 0 and no intra-match swing)
    // would just be noise.
    if (replay.gained > 0 || replay.lost > 0) result.breakdown.push(replay)
  }
  result.netPoints = result.pointsWon - result.pointsLost
  // Most volatile first (biggest total point movement), losses breaking ties so
  // a painful match outranks a lucky one of equal size.
  result.breakdown.sort((a, b) => b.gained + b.lost - (a.gained + a.lost) || b.lost - a.lost)
  return result
}

File: apps/mobile-flutter/parity/lib/fergie.dart
import 'scoring.dart' as scoring;

// Dart port of server/utils/analytics/fergie.ts (+ minuteValue from
// stats/insights.ts). Operates on JSON-shaped maps. The `oddsForOutcome`
// callback of the TS input cannot cross a frozen vector, so odds are always null
// here (the fergie vectors never score on ODDS); everything else is deterministic.

const _unknownMinute = 1000000000; // 1e9

final _minuteRe = RegExp(r'(\d+)(?:[^\d]*\+(\d+))?');

num minuteValue(String? minute) {
  if (minute == null || minute.isEmpty) return _unknownMinute;
  final m = _minuteRe.firstMatch(minute);
  if (m == null) return _unknownMinute;
  return int.parse(m.group(1)!) * 100 + (m.group(2) != null ? int.parse(m.group(2)!) : 0);
}

bool isAddedTime(String? minute) =>
    minute != null && minute.contains('+') && (minuteValue(minute) / 100).floor() >= 90;

Map<String, dynamic> _emptyFergie() => {
      'matches': 0,
      'goals': 0,
      'netPoints': 0,
      'pointsWon': 0,
      'pointsLost': 0,
      'biggestGain': null,
      'biggestLoss': null,
      'breakdown': [],
    };

num _realPoints(Map m, Map rules, Map at) {
  final scores = scoring.scorePredictions({
    'actual': at,
    'predictions': m['field'],
    'rules': rules,
    'actualOutcomeOdds': null,
    'forceJoker': m['forceJoker'],
  });
  for (final s in scores.cast<Map>()) {
    if (s['id'] == m['predId']) return s['totalPoints'] as num;
  }
  return 0;
}

Map<String, dynamic>? _replayMatch(Map m, Map rules) {
  final goals = (m['goals'] as List).cast<Map>();
  final ordered = [
    for (var i = 0; i < goals.length; i++)
      {'g': goals[i], 'order': minuteValue(goals[i]['minute'] as String?), 'i': i},
  ];
  if (ordered.any((o) => (o['order'] as num) >= _unknownMinute)) return null;
  // TS sorts with the stable Array.prototype.sort; Dart's List.sort is not
  // stable, so feed order is carried explicitly as the terminal tie-break.
  ordered.sort((a, b) {
    final byOrder = (a['order'] as num).compareTo(b['order'] as num);
    return byOrder != 0 ? byOrder : (a['i'] as int).compareTo(b['i'] as int);
  });
  final sorted = ordered.map((o) => o['g'] as Map).toList();

  var home = 0, away = 0;
  num gained = 0, lost = 0;
  var added = 0;
  for (final g in sorted) {
    final before = {'home': home, 'away': away};
    if (g['side'] == 'HOME') {
      home += 1;
    } else {
      away += 1;
    }
    if (!isAddedTime(g['minute'] as String?)) continue;
    added += 1;
    final after = {'home': home, 'away': away};
    if (m['isKnockout'] == true && scoring.outcomeOf(before) == 'DRAW') continue;
    final delta = _realPoints(m, rules, after) - _realPoints(m, rules, before);
    if (delta > 0) {
      gained += delta;
    } else if (delta < 0) {
      lost += -delta;
    }
  }
  final actual = m['actual'] as Map;
  if (home != actual['home'] || away != actual['away']) return null;
  if (added == 0) return null;

  final pred = m['pred'] as Map;
  return {
    'home': m['home'],
    'away': m['away'],
    'homeCode': m['homeCode'],
    'awayCode': m['awayCode'],
    'predicted': '${pred['home']}-${pred['away']}',
    'actual': '${actual['home']}-${actual['away']}',
    'gained': gained,
    'lost': lost,
    'net': gained - lost,
    'isJoker': m['isJoker'],
  };
}

Map<String, dynamic> computeFergie(List matches, Map rules) {
  final result = _emptyFergie();
  final breakdown = <Map<String, dynamic>>[];
  for (final m in matches.cast<Map>()) {
    final replay = _replayMatch(m, rules);
    if (replay == null) continue;
    final gained = replay['gained'] as num, lost = replay['lost'] as num;
    result['matches'] = (result['matches'] as num) + 1;
    result['goals'] = (result['goals'] as num) +
        (m['goals'] as List).cast<Map>().where((g) => isAddedTime(g['minute'] as String?)).length;
    result['pointsWon'] = (result['pointsWon'] as num) + gained;
    result['pointsLost'] = (result['pointsLost'] as num) + lost;

    final bg = result['biggestGain'] as Map?;
    if (gained > 0 && (bg == null || gained > (bg['gained'] as num))) result['biggestGain'] = replay;
    final bl = result['biggestLoss'] as Map?;
    if (lost > 0 && (bl == null || lost > (bl['lost'] as num))) result['biggestLoss'] = replay;

    if (gained > 0 || lost > 0) breakdown.add({'r': replay, 'i': breakdown.length});
  }
  result['netPoints'] = (result['pointsWon'] as num) - (result['pointsLost'] as num);
  // Same stability gap as the goal replay: TS keeps insertion order for entries
  // level on both keys, so the push index is the terminal tie-break here too.
  breakdown.sort((ea, eb) {
    final a = ea['r'] as Map, b = eb['r'] as Map;
    final byMove = ((b['gained'] as num) + (b['lost'] as num)) - ((a['gained'] as num) + (a['lost'] as num));
    if (byMove != 0) return byMove > 0 ? 1 : -1;
    final byLost = (b['lost'] as num) - (a['lost'] as num);
    if (byLost != 0) return byLost > 0 ? 1 : -1;
    return (ea['i'] as int).compareTo(eb['i'] as int);
  });
  result['breakdown'] = breakdown.map((e) => e['r']).toList();
  return result;
}
